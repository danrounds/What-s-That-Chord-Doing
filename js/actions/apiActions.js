import { getReqUserScores, getReqHighScores, postReqAccount,
         putReqUserScores, putReqAccountPassword, deleteReqAccount }
from '../apiCalls';

const getScores = {
    getUserScores: (name, password) =>
        dispatch => {
            dispatch(getScores.getUserScoresPending());
            return getReqUserScores(name, password)
                .then(scores => dispatch(getScores.getUserScoresSuccess(scores)))
                .catch((e) => dispatch(getScores.getUserScoresFailure(e)));
        },

    GET_USER_SCORES_PENDING: 'GET_USER_SCORES_PENDING',
    getUserScoresPending: () => ({
        type: GET_USER_SCORES_PENDING
    }),

    GET_USER_SCORES_SUCCESS: 'GET_USER_SCORES_SUCCESS',
    getUserScoresSuccess: (scores) => ({
        type: GET_USER_SCORES_SUCCESS,
        scores
    }),

    GET_USER_SCORES_FAILURE: 'GET_USER_SCORES_FAILURE',
    getUserScoresFailure: (error) => ({
        type: GET_USER_SCORES_FAILURE,
        error
    }),
};

export const LOG_OFF = 'LOG_OFF';
export const logOff = () => ({
    type: LOG_OFF
});

const updateScores = {
    updateUserScores: (name, password, scores) =>
        dispatch => putReqUserScores(name, password, scores)
        .then(() => dispatch(updateScores.updateUserScoresSuccess(scores)))
        .catch((e) => dispatch(updateScores.updateUserScoresFailure(e))),

    UPDATE_USER_SCORES_SUCCESS: 'UPDATE_USER_SCORES_SUCCESS',
    updateUserScoresSuccess: () => ({
        type: UPDATE_USER_SCORES_SUCCESS
    }),

    UPDATE_USER_SCORES_FAILURE: 'UPDATE_USER_SCORES_FAILURE',
    updateUserScoresFailure: (error) => ({
        type: UPDATE_USER_SCORES_FAILURE,
        error
    }),
};

const getHighScores_ = {
    getHighScores: (gameType) =>
        dispatch => {
            dispatch(getHighScores_.getHighScoresPending());
            return getReqHighScores(gameType)
                .then(highScores => dispatch(getHighScores_.getHighScoresSuccess(highScores)))
                .catch((e) => dispatch(getHighScores_.getHighScoresFailure(e)));
        },

    GET_HIGH_SCORES_PENDING: 'GET_HIGH_SCORES_PENDING',
    getHighScoresPending: () => ({
        type: GET_HIGH_SCORES_PENDING
    }),

    GET_HIGH_SCORES_SUCCESS: 'GET_HIGH_SCORES_SUCCESS',
    getHighScoresSuccess: (highScores) => ({
        type: GET_HIGH_SCORES_SUCCESS,
        highScores
    }),

    GET_HIGH_SCORES_FAILURE: 'GET_HIGH_SCORES_FAILURE',
    getHighScoresFailure: (error) => ({
        type: GET_HIGH_SCORES_FAILURE,
        error
    }),
}

const makeAccount = {
    makeUserAccount: (name, password) =>
        dispatch => postReqAccount(name, password)
        .then(() => dispatch(makeAccount.makeUserAccountSuccess()))
        .catch(() => dispatch(makeAccount.makeUserAccountFailure())),

    MAKE_USER_ACCOUNT_SUCCESS: 'MAKE_USER_ACCOUNT_SUCCESS',
    makeUserAccountSuccess: () => ({
        type: MAKE_USER_ACCOUNT_SUCCESS
    }),

    MAKE_USER_ACCOUNT_FAILURE: 'MAKE_USER_ACCOUNT_FAILURE',
    makeUserAccountFailure: (error) => ({
        type: MAKE_USER_ACCOUNT_FAILURE,
        error
    }),
};

const changePass = {
    changeUserPassword: (name, password) =>
        dispatch => putReqAccountPassword(name, password)
        .then(() => dispatch(changePass.changeUserPasswordSuccess()))
        .catch((e) => dispatch(changePass.changeUserPasswordFailure())),

    CHANGE_USER_PASSWORD_SUCCESS: 'CHANGE_USER_PASSWORD_SUCCESS',
    changeUserPasswordSuccess: () => ({
        type: CHANGE_USER_PASSWORD_SUCCESS
    }),

    CHANGE_USER_PASSWORD_FAILURE: 'CHANGE_USER_PASSWORD_FAILURE',
    changeUserPasswordFailure: (error) => ({
        type: CHANGE_USER_PASSWORD_FAILURE,
        error
    }),
};

///
///
// I've not bothered to write an action for account deletion, because that
// doesn't belong in the MVP client

// We end up with a bit more code by grouping associated actions together, but
// it at least tells us which goes with which
export const { getUserScores, GET_USER_SCORES_PENDING, getUserScoresPending,
               GET_USER_SCORES_SUCCESS, getUserScoresSuccess,
               GET_USER_SCORES_FAILURE, getUserScoresFailure } = getScores;

export const { updateUserScores, UPDATE_USER_SCORES_SUCCESS,
               updateUserScoresSuccess, UPDATE_USER_SCORES_FAILURE,
               updateUserScoresFailure } = updateScores;

export const { getHighScores, GET_HIGH_SCORES_PENDING, getHighScoresPending,
               GET_HIGH_SCORES_SUCCESS, getHighScoresSuccess,
               GET_HIGH_SCORES_FAILURE, getHighScoresFailure } = getHighScores_;

export const { makeUserAccount, MAKE_USER_ACCOUNT_SUCCESS,
               makeUserAccountSuccess, MAKE_USER_ACCOUNT_FAILURE,
               makeUserAccountFailure } = makeAccount;

export const { changeUserPassword, CHANGE_USER_PASSWORD_SUCCESS,
               changeUserPasswordSuccess, CHANGE_USER_PASSWORD_FAILURE,
               changeUserPasswordFailure } = changePass;
