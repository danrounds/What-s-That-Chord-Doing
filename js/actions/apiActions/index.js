export * from './actionsAccountAndPassword';
export * from './actionsLogInOff';
export * from './actionsScores';
