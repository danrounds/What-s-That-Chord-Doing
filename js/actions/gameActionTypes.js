export const SET_KEY_PRESS = 'SET_KEY_PRESS';
export const START_NEW_GAME = 'START_NEW_GAME';
export const GIVE_UP = 'GIVE_UP';
export const GET_NEXT_QUESTION = 'GET_NEXT_QUESTION';
export const MAKE_GUESS = 'MAKE_GUESS';
export const MARK_TURN_CORRECT = 'MARK_TURN_CORRECT';
