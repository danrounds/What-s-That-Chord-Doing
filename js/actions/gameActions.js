import chordGetter from '../musicLogic/';

export const SET_KEY_PRESS = 'SET_KEY_PRESS';
export const setKeyPress = (keyValue) => ({
    type: SET_KEY_PRESS,
    keyValue,
});

export const START_NEW_GAME = 'START_NEW_GAME';
export const startNewGame = (gameType, inversions) => {
    let {keyNameReadable, keyNameNotation, introChordSequence, chordSubset} =
            chordGetter.init(gameType, inversions);
    let {chordName, currentChordNumeral, bassNote, trebleNotes, accidentals} =
            chordGetter.getChord();
    let gameNumber = Math.random() * Math.pow(2,52);
    // ^ This is not a seed number; it's merely an [ostensibly] unique
    // identifier, that we'll use to decide whether components should update

    return {
        type: START_NEW_GAME,
        gameType,
        gameNumber,
        inversions,
        keyNameReadable,
        keyNameNotation,
        introChordSequence,
        chordSubset,
        chordName,
        currentChordNumeral,
        bassNote,
        trebleNotes,
        accidentals
    };
};
export const GIVE_UP = 'GIVE_UP';
export const giveUp = () => ({
    type: GIVE_UP,
});

export const GET_NEXT_QUESTION = 'GET_NEXT_QUESTION';
export const getNextQuestion = () => {
    let {chordName, currentChordNumeral, bassNote, trebleNotes, accidentals} =
            chordGetter.getChord();
    return {
        type: GET_NEXT_QUESTION,
        chordName,
        currentChordNumeral,
        bassNote,
        trebleNotes,
        accidentals,
    };
};

export const MAKE_GUESS = 'MAKE_GUESS';
export const makeGuess = (guess) => ({
    type: MAKE_GUESS,
    guess,
});

export const MARK_TURN_CORRECT = 'MARK_TURN_CORRECT';
export const markTurnCorrect = () => ({
    type: MARK_TURN_CORRECT,
});
