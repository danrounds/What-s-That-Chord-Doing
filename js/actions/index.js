export * from './uiActions';
export * from './gameActions';
export * from './apiActions';
