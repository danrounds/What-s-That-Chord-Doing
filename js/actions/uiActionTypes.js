export const UPDATE_LESSON_INDEX_DISPLAY = 'UPDATE_LESSON_INDEX_DISPLAY';
export const TOGGLE_KEYBOARD_SHORTCUT_DISPLAY = 'TOGGLE_KEYBOARD_SHORTCUT_DISPLAY';
