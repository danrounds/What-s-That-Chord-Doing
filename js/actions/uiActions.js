export const UPDATE_LESSON_INDEX_DISPLAY = 'UPDATE_LESSON_INDEX_DISPLAY';
export const updateLessonIndexDisplay = (keyValueToUpdate) => ({
    type: UPDATE_LESSON_INDEX_DISPLAY,
    keyValueToUpdate,
});

export const TOGGLE_KEYBOARD_SHORTCUT_DISPLAY = 'TOGGLE_KEYBOARD_SHORTCUT_DISPLAY';
export const toggleKeyboardShortcutDisplay = (displayBool) => {
    localStorage.setItem('@WTCD/displayKeyboardShortcuts', displayBool);
    return {
        type: TOGGLE_KEYBOARD_SHORTCUT_DISPLAY,
        displayBool,
    };
};
