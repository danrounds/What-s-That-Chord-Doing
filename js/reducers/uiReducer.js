import * as actions from '../actions/uiActions';

const initialState = {
    lessonIndexDisplay: { easy: false, novice: false, difficult: false, i: 0 },
    displayKeyboardShortcuts: JSON.parse(localStorage.getItem('@WTCD/displayKeyboardShortcuts')),
};

// Reducer for the lesson-index and keyboard shortcut display
export default (state=initialState, action) => {

    switch(action.type) {

    case actions.UPDATE_LESSON_INDEX_DISPLAY:
        return {
            lessonIndexDisplay: Object.assign({}, state.lessonIndexDisplay, action.keyValueToUpdate),
            displayKeyboardShortcuts: state.displayKeyboardShortcuts,
        };

    case actions.TOGGLE_KEYBOARD_SHORTCUT_DISPLAY:
        return {
            lessonIndexDisplay: Object.assign({}, state.lessonIndexDisplay),
            displayKeyboardShortcuts: action.displayBool,
        };

    default:
        return state;
    }
};

