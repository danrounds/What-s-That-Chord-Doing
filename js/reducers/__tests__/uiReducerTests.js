import uiReducer from '../uiReducer';
import * as actions from '../../actions/uiActions';

describe('gameReducer -- the reducer for our game actions', () => {
    let state;
    beforeEach(() => state = uiReducer(undefined, {}));

    it('should have a consistent initial state', () => {
        expect(state).toMatchSnapshot();
   });

    it('should update the state of our lesson display index', () => {
        state = uiReducer(state, actions.updateLessonIndexDisplay(3));
        expect(state).toMatchSnapshot();
    });

    it('should toggle display of keyboard shortcuts (to true)', () => {
        state = uiReducer(state, actions.toggleKeyboardShortcutDisplay(true));
        expect(state).toMatchSnapshot();
    });
    it('should toggle display of keyboard shortcuts (to false)', () => {
        state = uiReducer(state, actions.toggleKeyboardShortcutDisplay(false));
        expect(state).toMatchSnapshot();
    });
});
