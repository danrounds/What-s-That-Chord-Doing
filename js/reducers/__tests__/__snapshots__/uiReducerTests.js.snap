// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameReducer -- the reducer for our game actions should have a consistent initial state 1`] = `
Object {
  "displayKeyboardShortcuts": null,
  "lessonIndexDisplay": Object {
    "difficult": false,
    "easy": false,
    "i": 0,
    "novice": false,
  },
}
`;

exports[`gameReducer -- the reducer for our game actions should toggle display of keyboard shortcuts (to false) 1`] = `
Object {
  "displayKeyboardShortcuts": false,
  "lessonIndexDisplay": Object {
    "difficult": false,
    "easy": false,
    "i": 0,
    "novice": false,
  },
}
`;

exports[`gameReducer -- the reducer for our game actions should toggle display of keyboard shortcuts (to true) 1`] = `
Object {
  "displayKeyboardShortcuts": true,
  "lessonIndexDisplay": Object {
    "difficult": false,
    "easy": false,
    "i": 0,
    "novice": false,
  },
}
`;

exports[`gameReducer -- the reducer for our game actions should update the state of our lesson display index 1`] = `
Object {
  "displayKeyboardShortcuts": null,
  "lessonIndexDisplay": Object {
    "difficult": false,
    "easy": false,
    "i": 0,
    "novice": false,
  },
}
`;
